### DISCUSSION TITLE
getProviders doesn't return up to date values

### DISCUSSION THREAD

**Discussion by PedroCarreiras** [nextauthjs/next-auth#8528] [Help]
Hey! I am having some trouble with the getProviders function in my custom signin page.

My app currently allows for signin with google and github and I am trying to implement signin with gitlab. However, after I update src/app/apu/auth/[...nextauth]/route.ts and check http://172.29.92.3:3000/api/auth/providers to see if all the providers show up (which they do), the signin page still looks as in the following picture:

![image](https://github.com/nextauthjs/next-auth/assets/83094111/d08fb58b-0daa-4a37-917e-e441bfd68d26)

http://172.29.92.3:3000/api/auth/providers (formatted):
```json
{"google":{"id":"google","name":"Google","type":"oauth","signinUrl":"http://localhost:3000/api/auth/signin/google","callbackUrl":"http://localhost:3000/api/auth/callback/google"},
"github":{"id":"github","name":"GitHub","type":"oauth","signinUrl":"http://localhost:3000/api/auth/signin/github","callbackUrl":"http://localhost:3000/api/auth/callback/github"},
"gitlab":{"id":"gitlab","name":"GitLab","type":"oauth","signinUrl":"http://localhost:3000/api/auth/signin/gitlab","callbackUrl":"http://localhost:3000/api/auth/callback/gitlab"}}
```

src/app/apu/auth/[...nextauth]/route.ts currently looks like this

```ts
import NextAuth, { NextAuthOptions } from "next-auth";
import GithubProvider from "next-auth/providers/github"
import GoogleProvider from "next-auth/providers/google"
import GitlabProvider from "next-auth/providers/gitlab";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import { prisma } from '../../../client';

const google = GoogleProvider({
  clientId: process.env.ID_GOOGLE,
  clientSecret: process.env.SECRET_GOOGLE,
})

const github = GithubProvider({
  clientId: process.env.ID_GITHUB,
  clientSecret: process.env.SECRET_GITHUB,
})

const gitlab = GitlabProvider({
  clientId: process.env.ID_GITLAB,
  clientSecret: process.env.SECRET_GITLAB
})

export const authOptions: NextAuthOptions = {
  providers: [
    google,
    github,
    gitlab,
  ],
  adapter: PrismaAdapter(prisma),
  pages: {
    signIn: "/login",
  },
  callbacks: {
    async session({ session, user }) {
      if(user) {
        session.user.id = user.id;
      }
    
      return session;
    }
  },
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

And here is the signin page code
```ts
import styles from "./page.module.css";
import LogInButton from "@/components/loginbutton/loginbutton";
import { ClientSafeProvider, getProviders } from "next-auth/react";

export default async function LogIn() {
  const providers = await getProviders(); // doesn't work
  const res = await fetch("http://172.29.92.3:3000/api/auth/providers")
  const data: ClientSafeProvider[] = await res.json()

  const providerPaths : { [key: string]: string } = {
    'Google': '/providers/google.svg',
    'GitHub': '/providers/github.svg',
    'GitLab': '/providers/gitlab.svg',
  }

  return (
    <main className={styles.main}>
      <div className={styles.box}>
        {providers &&
          Object.values(providers).map((provider) => (
            <LogInButton key={provider.id} provider={provider} path={providerPaths[provider.name]} />
          ))}
      </div>
    </main>
  );
}
```

When I use data instead of providers in the map inside the return I get the following page. Something I noted is that now providers show up in the same order as the route file...

![image](https://github.com/nextauthjs/next-auth/assets/83094111/9fdeee10-a672-48be-895b-5bfafa993212)

In all of this, no errors are thrown and I couldn't find anything wrong according to documentation or other people's example apps.

Thanks in advance.

Bonus question(s): How can I get all of the provider icons to have the same size? (or is there a library or tool somewhere that will style the buttons for me? for example,  are the buttons used in the next auth default page a usable component like my LogInButton?)

Update: I have now tried to add discord as a provider and even with the previous solution the buttons don't seem to update. All providers still appear when accessing the route through the browser.

Update: Good news everyone! adding {cache: "no-store"} fixed the issue with the fetch call:

```ts
const res = await fetch("http://172.29.92.3:3000/api/auth/providers", {cache: "no-store"});
const data: ClientSafeProvider[] = await res.json();
```

This even confirms a suspicion I had that this was all somehow related to a cache somewhere. Closing and restarting everything (pc included) didn't fix the issue, so I decided not to include it so as not to induce error in a possible discussion. The only thing left is to find out how to make `getProviders()` do the same.

Updating the `fetchData()` function in `_utils.ts` as follows seems to fix the issue.

```ts
export async function fetchData<T = any>(
  path: string,
  __NEXTAUTH: AuthClientConfig,
  logger: LoggerInstance,
  { ctx, req = ctx?.req }: CtxOrReq = {}
): Promise<T | null> {
  const url = `${apiBaseUrl(__NEXTAUTH)}/${path}`
  try {
    const options: RequestInit = {
      headers: {
        "Content-Type": "application/json",
        ...(req?.headers?.cookie ? { cookie: req.headers.cookie } : {}),
      },
    }

    if (req?.body) {
      options.body = JSON.stringify(req.body)
      options.method = "POST"
      options.cache = 'no-cache'
    }

    const res = await fetch(url, options)
    const data = await res.json()
    if (!res.ok) throw data
    return Object.keys(data).length > 0 ? data : null // Return null if data empty
  } catch (error) {
    logger.error("CLIENT_FETCH_ERROR", { error: error as Error, url })
    return null
  }
}
```

However, I imagine this is just temporary solution and it wouldn't be advisable to keep it like this.

**Comment by cw-1988** [comment]
same issue here, manually navigating to api/auth/providers returns all 3 of my providers. using getProviders only returns one.
I'm running:
    "next": "14.0.3",
    "next-auth": "^4.24.5",
    
Deleteing the .next folder fixed the issue, so it's a caching one...
